pkg/logutil/merge_logger.go

```diff
@@ -0,0 +1,188 @@
+type MergeLogger struct {
+	*capnslog.PackageLogger
+
+	mu      sync.Mutex // protect statusm
+	statusm map[line]*status
+}
+
+func NewMergeLogger(logger *capnslog.PackageLogger) *MergeLogger {
+	l := &MergeLogger{
+		PackageLogger: logger,
+		statusm:       make(map[line]*status),
+	}
+	go l.outputLoop()
+	return l
+}
+
+func (l *MergeLogger) Info(mergePeriod time.Duration, entries ...interface{}) {
+	l.merge(mergePeriod, line{
+		level: capnslog.INFO,
+		str:   fmt.Sprint(entries...),
+	})
+}
+
+func (l *MergeLogger) Infof(mergePeriod time.Duration, format string, args ...interface{}) {
+	l.merge(mergePeriod, line{
+		level: capnslog.INFO,
+		str:   fmt.Sprintf(format, args...),
+	})
+}
+
+func (l *MergeLogger) Notice(mergePeriod time.Duration, entries ...interface{}) {
+	l.merge(mergePeriod, line{
+		level: capnslog.NOTICE,
+		str:   fmt.Sprint(entries...),
+	})
+}
+
+func (l *MergeLogger) Noticef(mergePeriod time.Duration, format string, args ...interface{}) {
+	l.merge(mergePeriod, line{
+		level: capnslog.NOTICE,
+		str:   fmt.Sprintf(format, args...),
+	})
+}
+
+func (l *MergeLogger) Warning(mergePeriod time.Duration, entries ...interface{}) {
+	l.merge(mergePeriod, line{
+		level: capnslog.WARNING,
+		str:   fmt.Sprint(entries...),
+	})
+}
+
+func (l *MergeLogger) Warningf(mergePeriod time.Duration, format string, args ...interface{}) {
+	l.merge(mergePeriod, line{
+		level: capnslog.WARNING,
+		str:   fmt.Sprintf(format, args...),
+	})
+}
+
+func (l *MergeLogger) Error(mergePeriod time.Duration, entries ...interface{}) {
+	l.merge(mergePeriod, line{
+		level: capnslog.ERROR,
+		str:   fmt.Sprint(entries...),
+	})
+}
+
+func (l *MergeLogger) Errorf(mergePeriod time.Duration, format string, args ...interface{}) {
+	l.merge(mergePeriod, line{
+		level: capnslog.ERROR,
+		str:   fmt.Sprintf(format, args...),
+	})
+}
+
+// Flush flushes summary log lines, and halts merging in progress.
+func (l *MergeLogger) Flush() {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	for ln, status := range l.statusm {
+		print(l.PackageLogger, ln, status)
+	}
+	l.statusm = make(map[line]*status)
+}
+
+func (l *MergeLogger) merge(period time.Duration, ln line) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	// increase count if the logger is merging the line
+	if status, ok := l.statusm[ln]; ok {
+		status.mu.Lock()
+		status.count++
+		status.mu.Unlock()
+		return
+	}
+
+	// print out the line at the first time
+	l.PackageLogger.Logf(ln.level, ln.str)
+	// initialize status of the line
+	l.statusm[ln] = &status{
+		period: period,
+		start:  time.Now(),
+	}
+}
+
+func (l *MergeLogger) outputLoop() {
+	for now := range time.Tick(time.Second) {
+		l.mu.Lock()
+		for ln, status := range l.statusm {
+			if !status.canPrint(now) {
+				continue
+			}
+			ok := print(l.PackageLogger, ln, status)
+			if !ok {
+				delete(l.statusm, ln)
+			}
+		}
+		l.mu.Unlock()
+	}
+}
+
+func print(logger *capnslog.PackageLogger, ln line, status *status) bool {
+	status.mu.Lock()
```

Comment:
You have a per-status lock, but AFAICT you are only ever touching status while holding the global MergeLogger mutex - am I wrong?